File: crates/daemon/src/runtime/managers/download.rs
use std::path::PathBuf;
use std::sync::Arc;

use engine::{Downloader, Engine};
use proto::download::{
    DownloadCancelledEvent, DownloadDoneEvent, DownloadErrorEvent, DownloadProgress,
    DownloadProgressEvent, DownloadSpec,
};

use super::job::{progress_event, settle, Cancellations, Runner, Spec};
use crate::runtime::EventHub;

pub struct DownloadManager {
    runner: Runner<String>,
}

impl DownloadManager {
    pub fn new(engine: Arc<Engine>, hub: Arc<EventHub>, cancellations: Cancellations) -> Self {
        DownloadManager {
            runner: Runner::new(engine, hub, cancellations),
        }
    }

    /// Start a download off-thread. Returns the job id.
    pub fn start(&self, spec: DownloadSpec) -> String {
        let job = Spec {
            id: spec.id.clone(),
            prefix: "download",
            key: None,
            progress: progress_event(|id, progress: &DownloadProgress| DownloadProgressEvent {
                id,
                progress: progress.clone(),
            }),
            done: settle(|id, path: PathBuf| DownloadDoneEvent { id, path }),
            cancelled: Some(settle(|id, ()| DownloadCancelledEvent { id })),
            error: settle(|id, e| DownloadErrorEvent {
                id,
                error: crate::runtime::engine_error(e),
            }),
        };

        self.runner
            .start(job, move |engine, reporter| {
                Box::pin(async move {
                    // Per chunk, so a large download stops promptly; its `.part`
                    // is discarded by the failure path rather than promoted.
                    let report = reporter.checked();
                    Downloader::new(Some(engine.cache()))
                        .fetch(
                            &spec.url,
                            &spec.destination,
                            spec.checksum.as_ref(),
                            &report,
                        )
                        .await
                        .map(|()| spec.destination.clone())
                })
            })
            .expect("a keyless job is never refused")
    }
}

File: frontend/src/api/download.ts
/**
 * The `download.start` channel: stream a URL to a daemon-local path through
 * the engine's verifying downloader. Resolves with the written path.
 */

import { call } from './core/ipc';
import { type JobRun, runJob } from './core/jobs';
import type { DownloadProgress, DownloadSpec } from './types/download';

export function start(
  spec: Omit<DownloadSpec, 'id'>,
  job: JobRun<DownloadProgress>,
): Promise<{ id: string; path: string }> {
  return runJob<{ id: string; path: string }, DownloadProgress>({
    ...job,
    topics: {
      progress: 'download.progress',
      done: 'download.done',
      error: 'download.error',
    },
    start: () => call('download.start', { ...spec, id: job.id }),
  });
}
